notary: Add admin rescue commands

johnsaigle opened the pull request:
- Add commands to change the status of delayed and blackholed messages
  in the Notary
- Change Notary commands to use MessageID consistently instead of a
  combination of pointers to MessagePublication and VAAHash
  
This PR is based on #4315 and will stay in draft until that PR is merged.

djb15 commented on node/pkg/adminrpc/adminserver.go:
```diff
@@ -1112,6 +1121,66 @@ func (s *nodePrivilegedService) ChainGovernorResetReleaseTimer(_ context.Context
+	}
+
+	return &nodev1.NotaryBlackholeDelayedMessageResponse{
+		Response: "Blackholed message",
```
Can you add in the `vaaId` string into the responses?

djb15 commented on node/pkg/notary/admincommands.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// RemoveBlackholedMsg removes a message from the blackholed list and deletes it from the database.
+func (n *Notary) RemoveBlackholedMsg(msgID string) error {
```
Should we add such messages that are removed from the blackholed list to the pending queue with 0 release time to ensure they are processed? I'm going to guess a reobservation would most likely lead to the message being blackholed again

johnsaigle replied:
It's added below.

djb15 commented on proto/node/v1/node.proto:
```diff
@@ -405,6 +407,42 @@ message ChainGovernorResetReleaseTimerResponse {
+  string response = 1;
+}
+
+message NotaryResetReleaseTimerRequest {
```
Missing the admin command for this message

djb15 commented on node/pkg/notary/notary.go:
```diff
@@ -329,7 +331,7 @@ func (n *Notary) blackhole(msg *common.MessagePublication) error {
 
 	// When a message is blackholed, it should be removed from the delayed list and database.
-	err := n.removeDelayed(msg)
+	err := n.removeDelayed(msg.MessageID())
```
Probably applies to the notary PR rather than this one, but should we attempt to remove from the delayed list + database before adding to the blackholed list? Just in case the removal fails we don't want the message in both places

djb15 reviewed (commented):
Just reviewed the latest commit

johnsaigle commented:
Now that the Notary has been merged, this is in my court to rebase and fix-up. In doing so I'll go back and respond to the comments

djb15 commented on node/pkg/notary/admincommands.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// ResetReleaseTimer resets the release timer for a delayed message.
+func (n *Notary) ResetReleaseTimer(msgID string, delayDays uint32) error {
```
Do we not also need the hook in `adminserver` to call into this?

johnsaigle replied:
Yeah I just realized that, added into the latest commit

johnsaigle commented on node/pkg/common/pendingmessage.go:
```diff
@@ -33,27 +33,24 @@ func (p PendingMessage) Compare(other PendingMessage) int {
+	// Compare with [PendingTransfer.Marshal].
+	// #nosec G115  -- int64 and uint64 have the same number of bytes, and Unix time won't be negative.
+	vaa.MustWrite(buf, binary.BigEndian, uint64(p.ReleaseTime.Unix()))
+	buf.Write(msgPubBz)
```
During testing the previous implementation was yielded a buffer that was 8 bytes shorter than expected. I think that the buffer was being overwritten by `MustWrite` rather than being appended to, so the timestamp was being lost.

`binary.Write` used by `vaa.MustWrite` apparently only works for fixed-size types 😨 

Compare [bytes.Write](https://pkg.go.dev/bytes#Buffer.Write) with [binary.Write](https://pkg.go.dev/encoding/binary#Write).

djb15 replied:
Thanks for the detailed explanation! Agree it would be nice review further as part of future work

johnsaigle commented:
@djb15 The latest changes now include a new `script/` under scripts. I ran it against `tilt` locally. It should cover the process of moving messages back and forth between the delayed and blackholed lists, as well as resetting the release timer.

In the process I also added some utility commands to inject synthetic messages for testing. These are only accessible in "unsafe dev mode". There's also an ability for admins to list the messages in both lists and introspect the release time for delayed messages.

It turns out this getting this to work involved modifying a few more files but the changes should be minimal.

Interestingly, I also noticed and fixed an existing bug in the admin service runnable. The Governor is supposed to be optional, but it was listed as a strict dependency for the admin service command. If it was disabled, the current admin server could not be launched, which would have prevented other admin actions like reobservations!

djb15 reviewed (approved):
Great script, thanks for taking the time to put this test together! Also tested myself and worked as expected. And nice catch on the strict dependency!
